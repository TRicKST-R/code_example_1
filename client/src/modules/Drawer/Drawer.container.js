import { connect } from 'react-redux'
import Drawer, { drawerWidth } from './Drawer.view'
import { drawerActions, signIn } from './Drawer.state'
import { bindActionCreators } from 'redux'
import { push } from 'react-router-redux'
import { signOut } from '../../redux/sessionState'
import { actions } from '../Dialog/DialogRoot/DialogRoot.state'
import { changeUser, removeUser } from '../../root/localUsersReducer'

export { drawerWidth }
export default connect(
  ({ drawerState, session, localUsers }) => ({
    drawerState,
    session,
    localUsers
  }),
  dispatch => ({
    drawerActions: bindActionCreators(drawerActions, dispatch),
    push: bindActionCreators(push, dispatch),
    signOut: bindActionCreators(signOut, dispatch),
    showDialog: bindActionCreators(actions.showDialog, dispatch),
    signIn: bindActionCreators(signIn, dispatch),
    changeUser: bindActionCreators(changeUser, dispatch),
    removeUser: bindActionCreators(removeUser, dispatch)
  })
)(Drawer)
