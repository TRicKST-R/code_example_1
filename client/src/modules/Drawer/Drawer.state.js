import createReducer from '../../utils/createReducer'
import { SIGN_IN } from '../../redux/actionTypes'

export const DrawerState = {
  opened: false,
  showedHiddenBlock: false
}

const TOGGLE_DRAWER = 'TOGGLE_DRAWER'
const TOGGLE_HIDDEN_BLOCK = 'TOGGLE_HIDDEN_BLOCK'

const drawerReducer = {
  [TOGGLE_DRAWER]: state => ({ ...state, opened: !state.opened }),
  [TOGGLE_HIDDEN_BLOCK]: state => ({
    ...state,
    showedHiddenBlock: !state.showedHiddenBlock
  })
}

export const toggleDrawer = () => ({
  type: TOGGLE_DRAWER
})

const toggleHiddenBlock = () => ({
  type: TOGGLE_HIDDEN_BLOCK
})

export const signIn = ({ email, password, full }) => ({
  type: SIGN_IN,
  payload: { email, password, full }
})

export const drawerActions = {
  toggleDrawer,
  toggleHiddenBlock
}

export const drawerState = createReducer(drawerReducer, DrawerState)
