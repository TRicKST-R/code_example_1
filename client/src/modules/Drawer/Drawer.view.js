/* eslint-disable flowtype/require-valid-file-annotation */

import React, { Component } from 'react'
import Header from './Components/DrawerHeader'
// import HiddenMenu from './Components/DrawerHiddenMenu'
import CenterMenu from './Components/DrawerCenterMenu'
import BottomMenu from './Components/DrawerBottomMenu'
import SignList from './Components/DrawerSignList'
import { withStyles } from 'material-ui/styles'
import Drawer from 'material-ui/Drawer'
import Hidden from 'material-ui/Hidden'
import { styles } from './theme'
import ExitToAppIcon from 'material-ui-icons/ExitToApp'
import SettingsIcon from 'material-ui-icons/Settings'
import ShoppingCart from 'material-ui-icons/ShoppingCart'
import Pages from 'material-ui-icons/Pages'
import People from 'material-ui-icons/People'
import Equalizer from 'material-ui-icons/Equalizer'
import AttachMoney from 'material-ui-icons/AttachMoney'
import Folder from 'material-ui-icons/Folder'
import School from 'material-ui-icons/School'
import Dashboard from 'material-ui-icons/Dashboard'
import CompareArrows from 'material-ui-icons/CompareArrows'
import Domain from 'material-ui-icons/Domain'

export class ResponsiveDrawer extends Component {
  constructor (props) {
    super(props)
    const { push, signOut } = props
    this.menuItems = {
      center: [
        {
          label: 'Dashboard',
          Icon: Dashboard,
          allowTo: {
            BREWERY: true,
            WHOLESALER: false
          },
          onClick: () => push('/dashboard')
        },
        {
          label: 'Users',
          Icon: People,
          allowTo: {
            BREWERY: false,
            WHOLESALER: false
          },
          onClick: () => push('/users')
        },
        {
          label: 'Products',
          Icon: Pages,
          allowTo: {
            BREWERY: true,
            WHOLESALER: true
          },
          onClick: () => push('/products')
        },
        {
          label: 'Transfers',
          Icon: CompareArrows,
          allowTo: {
            BREWERY: true,
            WHOLESALER: false
          },
          onClick: () => push('/transfers')
        },
        {
          label: 'Breweries',
          Icon: Domain,
          allowTo: {
            BREWERY: false,
            WHOLESALER: true
          },
          onClick: () => push('/breweries')
        },
        {
          label: 'Wholesalers',
          Icon: Domain,
          allowTo: {
            BREWERY: true,
            WHOLESALER: false
          },
          onClick: () => push('/wholesalers')
        },
        {
          label: 'Splits',
          Icon: Equalizer,
          allowTo: {
            BREWERY: true,
            WHOLESALER: false
          },
          onClick: () => push('/splits')
        },
        {
          label: 'Transactions',
          Icon: AttachMoney,
          allowTo: {
            BREWERY: false,
            WHOLESALER: true
          },
          onClick: () => push('/transactions')
        },
        {
          label: 'Orders',
          Icon: ShoppingCart,
          allowTo: {
            BREWERY: true,
            WHOLESALER: true
          },
          onClick: () => push('/orders')
        }, /*        {          label: 'Allocations',          Icon: LocalOffer,
          allowTo: {
            BREWERY: true,
            WHOLESALER: true
          },
          onClick: () => push('/allocations')
        }, */
        {
          label: 'Files',
          Icon: Folder,
          allowTo: {
            BREWERY: true,
            WHOLESALER: true
          },
          onClick: () => push('/files')
        },
        {
          label: 'User Guidelines',
          Icon: School,
          allowTo: {
            BREWERY: false,
            WHOLESALER: true
          },
          onClick: () => push('/guidelines')
        }
      ],
      bottom: [
        {
          label: 'Sign out',
          Icon: ExitToAppIcon,
          onClick: () => signOut()
        }
      ],
      hidden: [
        {
          label: 'Settings',
          Icon: SettingsIcon,
          onClick: () => push('/settings')
        }
      ]
    }
  }

  render () {
    // Access control
    let centerItemsToShow = []
    if (this.props.session.role !== undefined) {
      centerItemsToShow = this.menuItems.center.filter(
        ({ allowTo }) => allowTo[this.props.session.role]
      )
    }

    const {
      classes,
      drawerActions,
      drawerState,
      showDialog,
      signIn,
      changeUser,
      removeUser,
      session: { name = '', email = '' },
      localUsers
    } = this.props
    const { toggleDrawer, toggleHiddenBlock } = drawerActions
    const { showedHiddenBlock, opened } = drawerState

    const drawer = (
      <div className={classes.flexWrapper}>
        <Header
          toggleHiddenBlock={toggleHiddenBlock}
          primary={name}
          secondary={email}
        />

        {showedHiddenBlock ? (
          <SignList
            showDialog={showDialog}
            signIn={signIn}
            changeUser={changeUser}
            removeUser={removeUser}
            toggleHiddenBlock={toggleHiddenBlock}
            userList={localUsers.otherUsers}
          />
        ) : null}

        {/* showedHiddenBlock ? (
          <HiddenMenu
            opened={opened}
            toggleDrawer={toggleDrawer}
            items={this.menuItems.hidden}
          />
        ) : null */}

        <CenterMenu
          opened={opened}
          toggleDrawer={toggleDrawer}
          items={centerItemsToShow}
        />

        <BottomMenu
          opened={opened}
          toggleDrawer={toggleDrawer}
          items={this.menuItems.bottom}
        />
      </div>
    )

    return (
      <div>
        <Hidden>
          <Drawer
            variant='temporary'
            open={opened}
            classes={{
              paper: opened ? classes.drawerPaperDis : classes.drawerPaper
            }}
            onRequestClose={() => {
              if (showedHiddenBlock) {
                toggleHiddenBlock()
              }
              return toggleDrawer()
            }}
            ModalProps={{
              onBackdropClick: toggleDrawer
            }}
          >
            {drawer}
          </Drawer>
        </Hidden>
        <Hidden mdDown implementation='css'>
          <div className={opened ? classes.insideHide : classes.inside} />
          <Drawer
            variant='persistent'
            open={!opened}
            classes={{
              paper: opened ? classes.drawerPaperDis : classes.drawerPaper
            }}
            ModalProps={{
              onBackdropClick: toggleDrawer
            }}
          >
            {drawer}
          </Drawer>
        </Hidden>
      </div>
    )
  }
}

export default withStyles(styles, { withTheme: true })(ResponsiveDrawer)
