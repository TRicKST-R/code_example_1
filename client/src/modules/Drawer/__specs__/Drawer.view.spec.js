import React from 'react'
import { shallow } from 'enzyme'
import { ResponsiveDrawer } from '../Drawer.view'

describe('<ResponsiveDrawer />', () => {
  let toggleDrawer,
    toggleHiddenBlock,
    opened,
    wrapper,
    classes,
    disabled,
    drawerActions,
    showedHiddenBlock
  beforeEach(() => {
    classes = {
      navIconHide: 'navIconHide',
      drawerHeader: 'drawerHeader',
      drawerPaper: 'drawerPaper',
      drawerPaperDis: 'drawerPaperDis',
      flexCenter: 'flexCenter',
      flexWrapper: 'flexWrapper',
      inside: 'inside'
    }
    opened = true
    disabled = false
    drawerActions = {
      toggleDrawer: jest.fn(),
      toggleHiddenBlock: jest.fn()
    }
    showedHiddenBlock = true
    wrapper = shallow(
      <ResponsiveDrawer
        classes={classes}
        opened={opened}
        disabled={disabled}
        drawerActions={drawerActions}
        toggleHiddenBlock={toggleHiddenBlock}
        toggleDrawer={toggleDrawer}
        showedHiddenBlock={showedHiddenBlock}
      />
    )
  })
  it('should renders as expected', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should handle toggleHiddenBlock and toggleDrawer', () => {
    wrapper
      .childAt(0)
      .childAt(0)
      .props()
      .onRequestClose()
    expect(drawerActions.toggleHiddenBlock.mock.calls.length).toEqual(1)
    wrapper.setProps({ showedHiddenBlock: false })
    wrapper
      .childAt(0)
      .childAt(0)
      .props()
      .onRequestClose()
    expect(drawerActions.toggleDrawer.mock.calls.length).toEqual(2)
  })

  it('should render div className inside or insideHide, relates on props', () => {
    expect(
      wrapper
        .childAt(1)
        .childAt(0)
        .props().className
    ).toEqual(classes.inside)
    wrapper.setProps({ disabled: true })
    expect(
      wrapper
        .childAt(1)
        .childAt(0)
        .props().className
    ).toEqual(classes.insideHide)
  })

  it('should render div className drawerPaper or drawerPaperDis, relates on props', () => {
    expect(
      wrapper
        .childAt(1)
        .childAt(1)
        .props().classes
    ).toMatchObject({
      paper: classes.drawerPaper
    })
    wrapper.setProps({ disabled: true })
    expect(
      wrapper
        .childAt(1)
        .childAt(1)
        .props().classes
    ).toMatchObject({
      paper: classes.drawerPaperDis
    })
  })
})
