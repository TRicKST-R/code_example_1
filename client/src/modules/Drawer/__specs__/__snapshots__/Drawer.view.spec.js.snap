// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResponsiveDrawer /> should renders as expected 1`] = `
<div>
  <Hidden
    implementation="js"
    lgDown={false}
    lgUp={false}
    mdDown={false}
    mdUp={false}
    smDown={false}
    smUp={false}
    xlDown={false}
    xlUp={false}
    xsDown={false}
    xsUp={false}
  >
    <withStyles(Drawer)
      classes={
        Object {
          "paper": "drawerPaper",
        }
      }
      onRequestClose={[Function]}
      open={true}
      type="temporary"
    >
      <div
        className="flexWrapper"
      >
        <withStyles(HeaderBlock)
          primary="Name SurnameName Surname"
          secondary="surname@gmail.comsurname@gmail.com"
          toggleHiddenBlock={[Function]}
        />
        <withStyles(HiddenBlock)
          items={
            Array [
              Object {
                "Icon": [Function],
                "label": "Settings",
                "onClick": [Function],
              },
            ]
          }
          opened={true}
          toggleDrawer={[Function]}
        />
        <withStyles(CenterBlock)
          items={
            Array [
              Object {
                "Icon": [Function],
                "label": "Home",
                "onClick": [Function],
              },
              Object {
                "Icon": [Function],
                "label": "Demo app",
                "onClick": [Function],
              },
              Object {
                "Icon": [Function],
                "label": "About",
                "onClick": [Function],
              },
              Object {
                "Icon": [Function],
                "label": "Dialog",
                "onClick": [Function],
              },
            ]
          }
          opened={true}
          toggleDrawer={[Function]}
        />
        <withStyles(BottomBlock)
          items={
            Array [
              Object {
                "Icon": [Function],
                "label": "Sign out",
                "onClick": [Function],
              },
            ]
          }
          opened={true}
          toggleDrawer={[Function]}
        />
      </div>
    </withStyles(Drawer)>
  </Hidden>
  <Hidden
    implementation="css"
    lgDown={false}
    lgUp={false}
    mdDown={true}
    mdUp={false}
    smDown={false}
    smUp={false}
    xlDown={false}
    xlUp={false}
    xsDown={false}
    xsUp={false}
  >
    <div
      className="inside"
    />
    <withStyles(Drawer)
      classes={
        Object {
          "paper": "drawerPaper",
        }
      }
      open={true}
      type="persistent"
    >
      <div
        className="flexWrapper"
      >
        <withStyles(HeaderBlock)
          primary="Name SurnameName Surname"
          secondary="surname@gmail.comsurname@gmail.com"
          toggleHiddenBlock={[Function]}
        />
        <withStyles(HiddenBlock)
          items={
            Array [
              Object {
                "Icon": [Function],
                "label": "Settings",
                "onClick": [Function],
              },
            ]
          }
          opened={true}
          toggleDrawer={[Function]}
        />
        <withStyles(CenterBlock)
          items={
            Array [
              Object {
                "Icon": [Function],
                "label": "Home",
                "onClick": [Function],
              },
              Object {
                "Icon": [Function],
                "label": "Demo app",
                "onClick": [Function],
              },
              Object {
                "Icon": [Function],
                "label": "About",
                "onClick": [Function],
              },
              Object {
                "Icon": [Function],
                "label": "Dialog",
                "onClick": [Function],
              },
            ]
          }
          opened={true}
          toggleDrawer={[Function]}
        />
        <withStyles(BottomBlock)
          items={
            Array [
              Object {
                "Icon": [Function],
                "label": "Sign out",
                "onClick": [Function],
              },
            ]
          }
          opened={true}
          toggleDrawer={[Function]}
        />
      </div>
    </withStyles(Drawer)>
  </Hidden>
</div>
`;
